runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,435 @@
+        InputT, OutputT, RestrictionT, TrackerT extends RestrictionTracker<RestrictionT>>
+    extends PTransform<PCollection<InputT>, PCollection<OutputT>> {
+  private final DoFn<InputT, OutputT> fn;
+  private final DoFnSignature signature;
+
+  /**
+   * Creates the transform for the given original {@link ParDo} and {@link DoFn}.
+   *
+   * @param fn The splittable {@link DoFn} inside the original {@link ParDo} transform.
+   */
+  public SplittableParDo(DoFn<InputT, OutputT> fn) {
+    checkNotNull(fn, "fn must not be null");
+    this.fn = fn;
+    this.signature = DoFnSignatures.INSTANCE.getOrParseSignature(fn.getClass());
+    checkArgument(signature.processElement().isSplittable(), "fn must be a splittable DoFn");
+  }
+
+  @Override
+  public PCollection<OutputT> apply(PCollection<InputT> input) {
+    PCollection.IsBounded isFnBounded = signature.isBounded();
+    Coder<RestrictionT> restrictionCoder =
+        DoFnInvokers.INSTANCE.newByteBuddyInvoker(fn).invokeGetRestrictionCoder();
+    Coder<ElementRestriction<InputT, RestrictionT>> splitCoder =
+        ElementRestrictionCoder.of(input.getCoder(), restrictionCoder);
+
+    PCollection<KeyedWorkItem<String, ElementRestriction<InputT, RestrictionT>>> keyedWorkItems =
+        input
+            .apply(
+                "Pair with initial restriction",
+                ParDo.of(new PairWithRestrictionFn<InputT, OutputT, RestrictionT>(fn)))
+            .setCoder(splitCoder)
+            .apply("Split restriction", ParDo.of(new SplitRestrictionFn<InputT, RestrictionT>(fn)))
+            .setCoder(splitCoder)
+            .apply(
+                "Assign unique key",
+                WithKeys.of(new RandomUniqueKeyFn<ElementRestriction<InputT, RestrictionT>>()))
+            .apply(
+                "Group by key",
+                new GBKIntoKeyedWorkItems<String, ElementRestriction<InputT, RestrictionT>>());
+    checkArgument(
+        keyedWorkItems.getWindowingStrategy().getWindowFn() instanceof GlobalWindows,
+        "GBKIntoKeyedWorkItems must produce a globally windowed collection, "
+            + "but windowing strategy was: %s",
+        keyedWorkItems.getWindowingStrategy());
+    return keyedWorkItems
+        .apply(
+            "Process",
+            ParDo.of(
+                new ProcessFn<InputT, OutputT, RestrictionT, TrackerT>(
+                    fn,
+                    input.getCoder(),
+                    input.getWindowingStrategy().getWindowFn().windowCoder())))
+        .setIsBoundedInternal(input.isBounded().and(isFnBounded))
+        .setWindowingStrategyInternal(input.getWindowingStrategy());
+  }
+
+  /**
+   * Assigns a random unique key to each element of the input collection, so that the output
+   * collection is effectively the same elements as input, but the per-key state and timers are now
+   * effectively per-element.
+   */
+  private static class RandomUniqueKeyFn<T> implements SerializableFunction<T, String> {
```

Comment:
Should this use the new transform to assign keys without using UUID? Should that be a TODO?